Objects/typevarobject.c

```diff
@@ -0,0 +1,1724 @@
+\n\
+  class C(Generic[*Ts]): ...\n\
+\n\
+Then we can parameterize that class with an arbitrary number of type\n\
+arguments:\n\
+\n\
+  C[int]       # Fine\n\
+  C[int, str]  # Also fine\n\
+  C[()]        # Even this is fine\n\
+\n\
+For more details, see PEP 646.\n\
+\n\
+Note that only TypeVarTuples defined in global scope can be pickled.\n\
+");
+
+PyType_Slot typevartuple_slots[] = {
+    {Py_tp_doc, (void *)typevartuple_doc},
+    {Py_tp_members, typevartuple_members},
+    {Py_tp_methods, typevartuple_methods},
+    {Py_tp_new, typevartuple},
+    {Py_tp_iter, typevartuple_iter},
+    {Py_tp_repr, typevartuple_repr},
+    {Py_tp_dealloc, typevartuple_dealloc},
+    {Py_tp_alloc, PyType_GenericAlloc},
+    {Py_tp_free, PyObject_GC_Del},
+    {Py_tp_traverse, typevartuple_traverse},
+    {Py_tp_clear, typevartuple_clear},
+    {0, 0},
+};
+
+PyType_Spec typevartuple_spec = {
+    .name = "typing.TypeVarTuple",
+    .basicsize = sizeof(typevartupleobject),
+    .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_IMMUTABLETYPE | Py_TPFLAGS_MANAGED_DICT
+        | Py_TPFLAGS_HAVE_GC,
+    .slots = typevartuple_slots,
+};
+
+PyObject *
+_Py_make_typevar(const char *name, PyObject *evaluate_bound, PyObject *evaluate_constraints)
+{
+    return (PyObject *)typevar_alloc(name, NULL, evaluate_bound, NULL, evaluate_constraints,
+                                     false, false, true, NULL);
+}
+
+PyObject *
+_Py_make_paramspec(PyThreadState *unused, PyObject *v)
+{
+    assert(PyUnicode_Check(v));
+    return (PyObject *)paramspec_alloc(PyUnicode_AsUTF8(v), NULL, false, false, true, NULL);
+}
+
+PyObject *
+_Py_make_typevartuple(PyThreadState *unused, PyObject *v)
+{
+    assert(PyUnicode_Check(v));
+    return (PyObject *)typevartuple_alloc(PyUnicode_AsUTF8(v), NULL);
+}
+
+static void
+typealias_dealloc(PyObject *self)
+{
+    PyTypeObject *tp = Py_TYPE(self);
+    _PyObject_GC_UNTRACK(self);
+    typealiasobject *ta = (typealiasobject *)self;
+    free((void *)ta->name);
+    Py_XDECREF(ta->type_params);
+    Py_XDECREF(ta->compute_value);
+    Py_XDECREF(ta->value);
+    Py_TYPE(self)->tp_free(self);
+    Py_DECREF(tp);
+}
+
+static PyObject *
+typealias_evaluate(typealiasobject *ta)
+{
+    if (ta->value != NULL) {
+        return Py_NewRef(ta->value);
+    }
+    PyObject *result = PyObject_CallNoArgs(ta->compute_value);
+    if (result == NULL) {
+        return NULL;
+    }
+    ta->value = Py_NewRef(result);
+    return result;
+}
+
+static PyObject *
+typealias_repr(PyObject *self)
+{
+    Py_ssize_t res = Py_ReprEnter(self);
+    if (res > 0) {
+        return PyUnicode_FromString("...");
+    }
+    else if (res < 0) {
+        return NULL;
+    }
+
+    typealiasobject *ta = (typealiasobject *)self;
+    PyObject *value_repr = NULL;
+    PyObject *value = typealias_evaluate(ta);
+    if (value == NULL) {
+        PyErr_Clear();
+        value_repr = PyUnicode_FromString("<evaluation failed>");
+    }
```

Comment:
I think so. Consider a type alias that depends on a name imported in `if TYPE_CHECKING`: evaluating the value would fail, but it would still be nice to be able to show a repr().

I actually might want to simplify the repr() of type aliases. The current semantics is from a discussion with Guido at the sprints, who thought it was important to show the value of the alias, but I don't like how complicated generic aliases can now look:

```
>>> from typing import Literal
>>> type BinOp = Literal["+", "-"]
>>> type LeftParen = Literal["("]
>>> type RightParen = Literal[")"]
>>> type SimpleExpr = int | Parenthesized
>>> type Parenthesized = tuple[LeftParen, Expr, RightParen]
>>> type Expr = SimpleExpr | tuple[SimpleExpr, BinOp, Expr]
>>> 
>>> Expr
<type alias Expr: <type alias SimpleExpr: int | <type alias Parenthesized: tuple[<type alias LeftParen: typing.Literal['(']>, ..., <type alias RightParen: typing.Literal[')']>]>> | tuple[<type alias SimpleExpr: int | <type alias Parenthesized: tuple[<type alias LeftParen: typing.Literal['(']>, ..., <type alias RightParen: typing.Literal[')']>]>>, <type alias BinOp: typing.Literal['+', '-']>, ...]>
```

Reply:
A repr that is just the name of the type alias would look _much_ nicer when it shows up in help() output as a type annotation.

Reply:
@gvanrossum fyi, I'm planning to go back to having the repr() be just the name of the alias based on this discussion.

Reply:
New behavior for the above example:

```
>>> Expr
Expr
>>> Expr.__value__
SimpleExpr | tuple[SimpleExpr, BinOp, Expr]
```